tensorflow/core/ops/math_ops.cc

```diff
@@ -1447,4 +1447,56 @@ b: Another tensor, of same type and shape as `a`.
 product: Pairwise cross product of the vectors in `a` and `b`.
 )doc");
 
+// --------------------------------------------------------------------------
+
+REGISTER_OP("Cumsum")
+    .Input("x: T")
+    .Input("axis: int32")
+    .Attr("reverse: bool")
+    .Output("out: T")
+    .Attr("T: numbertype")
+    .Doc(R"doc(
+Compute the cumulative sum of the tensor `x` along `axis`.
+
+The output `out` at any given index i is equal to the sum of all
+elements `x_j` of `x` with j <= i.
+
+By setting the `reverse` operation to `True`, the sum is performed in the
+reverse order. In contrast to using `tf.reverse`, this avoids copying the
+tensor.
+
+For example:
+
+```prettyprint
+# tensor 'x' is [1, 2, 3, 4, 5]
+tf.cumsum(x)               ==> [1, 3, 6, 10, 15]
+tf.cumsum(x, reverse=True) ==> [5, 9, 12, 14, 15]
```

Comment:
I don't think it would require an entirely separate Op- I'm just suggesting splitting up the `reverse` parameter into two distinct reversal options. Right now it indicates that it should both reverse the order of the summation (starting from the back) as well as reverse the order of the output tensor.

My thought is to add two more cases in [this section](https://github.com/tensorflow/tensorflow/pull/2711/files#diff-b6fb1bb5fdce94575b577ae6c5082e2fR32) (one each for just `reverse_sum == true` and just `reverse_order == true`) as well as changing `reverse` into those two separate parameters. The current `if (reverse)` would be converted to `if (reverse_sum && reverse_order)`